File: 2025/01/solution.py
lines = []

with open("input.txt", "r") as f:
    lines = f.readlines()


def part1():
    idx = 50
    solution = 0

    for line in lines:
        move = line[1:].strip()
        move = int(move)

        if line.startswith("L"):
            move *= -1

        idx += move
        idx %= 100

        if idx == 0:
            solution += 1

    return solution


def part2():
    idx = 50
    solution = 0

    for line in lines:
        move = line[1:].strip()
        move = int(move)

        step = 1
        if line.startswith("L"):
            step = -1

        while move != 0:
            idx += step
            move -= 1

            idx = idx % 100

            if idx == 0:
                solution += 1

    return solution


if __name__ == "__main__":
    print(f"Solution part 1: {part1()}")
    print(f"Solution part 2: {part2()}")

File: 2025/01/solution.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {
	body, err := ioutil.ReadFile("input.txt")

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	lines := strings.Split(string(body), "\n")

	solution1 := 0
	solution2 := 0
	idx := 50

	for _, line := range lines {
		move, err := strconv.ParseInt(strings.TrimSpace(line[1:]), 10, 64)
		if err != nil {
			fmt.Printf("Can't convert string to number %s\n", line[1:])
			continue
		}

		step := 1
		if strings.HasPrefix(line, "L") {
			step = -1
		}

		for move != 0 {
			idx += step
			move -= 1

			idx %= 100

			if idx == 0 {
				solution2 += 1
			}
		}

		if idx == 0 {
			solution1 += 1
		}
	}

	fmt.Printf("Solution for part 1: %d\n", solution1)
	fmt.Printf("Solution for part 2: %d\n", solution2)
	os.Exit(0)
}
